server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateStep.java

```diff
@@ -0,0 +1,100 @@
+
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.config.Settings;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.server.computation.ReportQueue;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor;
+import org.sonar.server.computation.component.ProjectSettingsRepository;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.qualitygate.MutableQualityGateHolder;
+import org.sonar.server.computation.qualitygate.QualityGate;
+import org.sonar.server.computation.qualitygate.QualityGateService;
+
+import static org.sonar.server.computation.component.Component.Type.PROJECT;
+import static org.sonar.server.computation.component.DepthTraversalTypeAwareVisitor.Order.PRE_ORDER;
+
+/**
+ * This step retrieves the QualityGate for the current {@link ReportQueue.Item} and stores it in
+ * {@link MutableQualityGateHolder}.
+ */
+public class QualityGateStep implements ComputationStep {
+  private static final Logger LOGGER = Loggers.get(QualityGateStep.class);
+
+  private static final String PROPERTY_QUALITY_GATE = "sonar.qualitygate";
+  private static final QualityGate DEFAULT_QUALITY_GATE = new QualityGate("default Quality Gate", null);
+
+  private final TreeRootHolder treeRootHolder;
+  private final ProjectSettingsRepository projectSettingsRepository;
+  private final QualityGateService qualityGateService;
+  private final MutableQualityGateHolder qualityGateHolder;
+
+  public QualityGateStep(TreeRootHolder treeRootHolder, ProjectSettingsRepository projectSettingsRepository,
+    QualityGateService qualityGateService, MutableQualityGateHolder qualityGateHolder) {
+    this.treeRootHolder = treeRootHolder;
+    this.projectSettingsRepository = projectSettingsRepository;
+    this.qualityGateService = qualityGateService;
+    this.qualityGateHolder = qualityGateHolder;
+  }
+
+  @Override
+  public void execute() {
+    new DepthTraversalTypeAwareVisitor(PROJECT, PRE_ORDER) {
+      @Override
+      public void visitProject(Component project) {
+        executeForProject(project);
+      }
+    }.visit(treeRootHolder.getRoot());
+  }
+
+  private void executeForProject(Component project) {
+    String projectKey = project.getKey();
+    if (projectKey == null) {
+      throw new IllegalStateException("Key of project component must be set before invoking this step");
+    }
+
+    Settings settings = projectSettingsRepository.getProjectSettings(projectKey);
+    String qualityGateSetting = settings.getString(PROPERTY_QUALITY_GATE);
+
+    if (qualityGateSetting == null || StringUtils.isBlank(qualityGateSetting)) {
+      LOGGER.debug("No quality gate is configured for project " + projectKey);
+      qualityGateHolder.setQualityGate(DEFAULT_QUALITY_GATE);
+      return;
+    }
+
+    try {
+      long qualityGateId = Long.valueOf(qualityGateSetting.trim());
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use "Long.parseLong" for this string-to-long conversion. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2130)